src/NServiceBus.Transport.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -371,59 +372,44 @@ async Task Process(AsyncEventingBasicConsumer consumer, BasicDeliverEventArgs me
             }
             catch (Exception ex)
             {
-                Logger.Error(
-                    $"Failed to retrieve ID from poison message. Moving message to queue '{settings.ErrorQueue}'...",
-                    ex);
+                Logger.Error($"Failed to retrieve ID from poison message. Moving message to queue '{settings.ErrorQueue}'...", ex);
                 await MovePoisonMessage(consumer, message, settings.ErrorQueue, messageProcessingCancellationToken).ConfigureAwait(false);
 
                 return;
             }
 
-            var processed = false;
-            var errorHandled = false;
-            var numberOfDeliveryAttempts = 0;
+            var numberOfDeliveryAttempts = GetDeliveryAttempts(message, messageId);
```

Comment:
Could we change the logic to only track delivery attempts if a message has failed i.e. move this line to 392? This way we won't need as large cache. Successful attempts won't even touch the cache. The difference is we will be tracking failures, not attempts but otherwise it should be fine.

Reply:
This is already happening because of the early exit when the redelivered flag is false.

Reply:
Thinking about this more, I can see now how moving the call could have additional benefit even with the early exit.

>  The difference is we will be tracking failures, not attempts but otherwise it should be fine. I think that such a change would only require as to use k => 1 in line 447.

This I'm still not understanding, though. You seem to be implying that there is a difference between attempts and failures, but at the point of needing to populate the `ErrorContext` value, those seem to be identical to me. 

Moving the call seems fine to me, but I don't see how that changes the logic of the new method in any way.